src/emqx_client.erl

```diff
@@ -248,12 +269,55 @@ parse_subopt([{nl, false} | Opts], Result) ->
 parse_subopt([{qos, QoS} | Opts], Result) ->
     parse_subopt(Opts, Result#{qos := ?QOS_I(QoS)}).
 
+-spec(request(client(), topic(), payload(), qos() | [pubopt()])
+        -> ok | {ok, packet_id()} | {error, term()}).
+request(Client, Topic, Payload, QoS) when is_binary(Topic), is_atom(QoS) ->
+    request(Client, Topic, Payload, [{qos, ?QOS_I(QoS)}]);
+request(Client, Topic, Payload, QoS) when is_binary(Topic), ?IS_QOS(QoS) ->
+    request(Client, Topic, Payload, [{qos, QoS}]);
+request(Client, Topic, Payload, Opts) when is_binary(Topic), is_list(Opts) ->
+    request(Client, Topic, #{}, Payload, Opts).
+
+-spec(request(client(), topic(), properties(), payload(), [pubopt()])
+        -> ok | {ok, packet_id()} | {error, term()}).
+request(Client, Topic, Properties, Payload, Opts)
+    when is_binary(Topic), is_map(Properties), is_list(Opts) ->
+    ok = emqx_mqtt_props:validate(Properties),
+    Retain = proplists:get_bool(retain, Opts),
+    Shared = proplists:get_bool(shared, Opts),
+    TimeOut = proplists:get_value(timeout, Opts, 5),
+    QoS = ?QOS_I(proplists:get_value(qos, Opts, ?QOS_0)),
+    {ok, ResponseTopic} = sub_response_topic(Client, Shared, QoS, Topic),
+    ClientId = gen_statem:call(Client, client_id),
+    NewProperties = maps:merge(Properties, #{'Response-Topic' => ResponseTopic,
+                                             'Correlation-Data' => ClientId}),
+    publish(Client, #mqtt_msg{qos  = QoS,
+                              retain = Retain,
+                              topic = ResponseTopic,
+                              props = NewProperties,
+                              payload = iolist_to_binary(Payload)}),
+    receive_response(Client, ClientId, TimeOut).
+
+receive_response(Client, ClientId, TimeOut) ->
+    receive
+        {publish, Response} ->
+            {ok, Properties} = maps:find(properties, Response),
+            case maps:find('Correlation-Data', Properties) of
+                {ok, ClientId} ->
+                    maps:find(payload, Response);
+                _ ->
+                    receive_response(Client, ClientId, TimeOut)
+            end
+    after TimeOut ->
+            {timeout, <<"No Response">>}
```

Comment:
the spec is `{error, term()}` but it's `{timeout, binary()}` returned here.

Reply:
I didn't add spec for this `receive_response` function.

Reply:
no you didn't add spec for this function, but it is called by `request/5` and it has a spec.